Add Typing to `baseio`, `baserawio`, and `exampleio`

zm711 opened the pull request:
Replaces #1229 because that now has merge conflicts.

I know this has been discussed #1123. Sam is against, others are typically for. This is just a draft to think about (& correct or reject), but I think adding type hints to some places exposed to the user would be beneficial. By putting the hints into the base it propagates to all ios to minimize other typing needing to be added, although developers can override the typing in their specific io.

I know that neo has complicated inputs, but I think type hints work best to give a suggested appropriate input (And as a nod to Sam I didn't add type hints for the returns yet).

pep8speaks commented:
Hello @zm711! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`neo/io/baseio.py`](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/io/baseio.py):

> [Line 101:46](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/io/baseio.py#L101): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 116:24](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/io/baseio.py#L116): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 116:29](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/io/baseio.py#L116): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 116:30](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/io/baseio.py#L116): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals

* In the file [`neo/rawio/baserawio.py`](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py):

> [Line 137:39](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L137): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 137:40](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L137): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 477:48](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L477): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 477:80](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L477): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 477:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L477): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (146 > 99 characters)
> [Line 477:111](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L477): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 477:140](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L477): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 497:82](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L497): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 497:87](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L497): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 497:88](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L497): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 508:85](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L508): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 508:90](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L508): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 508:91](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L508): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 519:57](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L519): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 519:63](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L519): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 519:64](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L519): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 532:56](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L532): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 532:75](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L532): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 532:90](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L532): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 532:97](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L532): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 532:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L532): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (125 > 99 characters)
> [Line 532:114](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L532): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 532:121](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L532): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 533:49](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L533): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 533:56](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L533): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 533:88](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L533): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 533:95](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L533): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 533:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L533): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (136 > 99 characters)
> [Line 533:125](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L533): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 533:132](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L533): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 534:54](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L534): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 534:59](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L534): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 534:60](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L534): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 534:78](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L534): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 534:83](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L534): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 534:84](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L534): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 591:82](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L591): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 591:83](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L591): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 591:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L591): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 99 characters)
> [Line 591:111](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L591): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 591:116](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L591): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 591:117](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L591): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 592:63](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L592): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 592:68](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L592): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 592:69](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L592): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 592:99](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L592): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 592:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L592): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (144 > 99 characters)
> [Line 592:104](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L592): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 592:105](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L592): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 592:133](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L592): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 592:138](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L592): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 592:139](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L592): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 631:43](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L631): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 631:44](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L631): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 631:61](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L631): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 631:62](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L631): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 631:84](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L631): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 631:90](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L631): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 647:84](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L647): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 647:85](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L647): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 654:55](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L654): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 654:56](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L654): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 654:73](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L654): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 654:74](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L654): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 654:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L654): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 99 characters)
> [Line 654:101](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L654): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 654:102](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L654): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 655:47](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L655): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 655:54](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L655): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 655:73](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L655): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 655:78](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L655): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 655:79](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L655): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 660:86](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L660): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 661:62](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L661): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 675:38](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L675): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 675:44](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L675): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 675:63](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L675): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 675:92](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L675): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 678:54](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L678): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 678:73](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L678): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 678:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L678): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 99 characters)
> [Line 678:102](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L678): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 679:44](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L679): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 679:51](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L679): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 679:70](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L679): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 679:75](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L679): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 679:76](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L679): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 697:84](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L697): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 697:85](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L697): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 698:40](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L698): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 698:46](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L698): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 704:81](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L704): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 705:40](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L705): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 705:46](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L705): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 711:45](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L711): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 783:42](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L783): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 783:46](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L783): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 804:85](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L804): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 804:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L804): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 99 characters)
> [Line 804:103](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L804): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 805:78](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L805): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 823:71](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L823): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 823:91](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L823): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 832:74](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L832): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 832:94](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L832): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 832:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L832): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 840:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L840): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (137 > 99 characters)
> [Line 840:111](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L840): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 840:131](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L840): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 850:35](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L850): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 850:36](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/baserawio.py#L850): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals

* In the file [`neo/rawio/examplerawio.py`](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py):

> [Line 88:37](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L88): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 88:44](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L88): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 270:50](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L270): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 270:65](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L270): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 270:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L270): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 99 characters)
> [Line 271:79](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L271): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 271:84](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L271): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 271:90](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L271): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 311:39](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L311): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 311:54](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L311): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 311:79](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L311): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 318:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L318): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (137 > 99 characters)
> [Line 318:111](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L318): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 318:131](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L318): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 347:48](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L347): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 347:68](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L347): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 383:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L383): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (137 > 99 characters)
> [Line 383:111](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L383): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 383:131](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L383): [E227](https://duckduckgo.com/?q=pep8%20E227) missing whitespace around bitwise or shift operator
> [Line 414:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L414): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 99 characters)
> [Line 422:100](https://github.com/NeuralEnsemble/python-neo/blob/0279e44431ea4e63f5d3849ed4149b914dfd51e9/neo/rawio/examplerawio.py#L422): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 99 characters)

zm711 commented:
I'll fix formatting after I know whether this is supported or not.

JoeZiminski commented:
I would add my support to such a PR. Typing is fantastic for catching bugs, in general improving robustness. Since I started using type hints last year they have caught a lot of potential bugs and saved me a lot of trouble!

samuelgarcia commented:
Having the role of the old dinosaur is not so easy in a group but I am trying my best to maintain my public character.

zm711 commented:
> Having the role of the old dinosaur is not so easy in a group but I am trying my best to maintain my public character.

@samuelgarcia never change. Your example error messages here and in probeinterface are always the highlight of my proofreading.

alejoe91 commented:
@zm711 this looks good to me! 

@samuelgarcia can we merge?

zm711 commented:
Thanks @alejoe91. I have a feeling when there is typing in the PR title @samuelgarcia has a filter for his alerts to jeter dans la poubelle.

samuelgarcia commented:
@zm711 it is even worst. typing PR are triggering in me a brain state nearby freezing.
I always need to turn off the computer for at least one hour to recover totally.

apdavison commented on neo/io/baseio.py:
```diff
@@ -96,7 +98,7 @@ class BaseIO:
     mode = 'file'  # or 'fake' or 'dir' or 'database'
 
-    def __init__(self, filename=None, **kargs):
+    def __init__(self, filename: str | Path = None, **kargs):
```
The use of `|` in type hints was added in Python 3.10. Since we still support Python 3.8 and 3.9, this should be `Union[str, Path]`

apdavison commented on neo/io/baseio.py:
```diff
@@ -111,7 +113,7 @@ def __init__(self, filename=None, **kargs):
 
     ######## General read/write methods #######################
-    def read(self, lazy=False, **kargs):
+    def read(self, lazy: bool = False, **kargs):
```
It would be nice to have an annotation for the return type, but perhaps that's for a future PR

zm711 replied:
Sam is so against typing I was trying to make small steps to start to get everyone on board. I think doing returns would be better in a second PR just to slowly move toward full typing.

apdavison commented on neo/rawio/baserawio.py:
```diff
@@ -369,7 +370,7 @@ def block_count(self):
         return self.header['nb_block']
 
-    def segment_count(self, block_index):
+    def segment_count(self, block_index: int):
```
add annotation for the return type?

apdavison commented on neo/rawio/baserawio.py:
```diff
@@ -473,7 +474,11 @@ def channel_id_to_index(self, stream_index, channel_ids):
-    def _get_channel_indexes(self, stream_index, channel_indexes, channel_names, channel_ids):
+    def _get_channel_indexes(self,
+                             stream_index: int,
+                             channel_indexes: list[int] | None,
```
same comment as above concerning `|` vs `Union`.
Strictly speaking this should also use `List`, since the use of `list` in this way was only introduced in Python 3.8, but since we'll be dropping 3.8 support around April 2024, maybe we can ignore that.

apdavison commented on neo/rawio/baserawio.py:
```diff
@@ -587,8 +602,13 @@ def get_analogsignal_chunk(self, block_index=0, seg_index=0, i_start=None, i_sto
-                                    channel_indexes=None, channel_names=None, channel_ids=None):
+    def rescale_signal_raw_to_float(self,
+                                    raw_signal: np.ndarray,
+                                    dtype: np.dtype = 'float32',
```
shouldn't this be `Union[np.dtype, str]`, since `'float32'` is a string?

zm711 replied:
I thought np.dtype included "float32" since numpy accepts float32? But maybe I'm wrong. We don't accept any string (for example I can do "float100000") so I think the only floats accepted are numpy floats which are contained by the numpy.dtype.

apdavison replied:
I don't have a very deep understanding of type hints in Python, but my reasoning was based on this:
```
In [1]: import numpy as np

In [2]: x = np.arange(5)

In [3]: x.dtype
Out[3]: dtype('int64')

In [4]: isinstance(x.dtype, np.dtype)
Out[4]: True

In [5]: isinstance("int64", np.dtype)
Out[5]: False
```

zm711 replied:
That's a great point I meant more in the sense that you can add `int64` as the dtype in any np.function.

```
import numpy as np

>>> x = np.arange(5, dtype='int64')
>>> x.dtype
dtype('int64')

>>> x = np.arange(5, dtype='int2')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: data type 'int2' not understood
```

It won't accept any string, but only the strings allowable by the numpy.dtype constructor. So you're right in that it is a string as its type, but it's only the strings accepted by numpy. Maybe there is a better way to represent that, but I thought the most succinct (and generally understandable) way to do this would be to tell people to just use `np.type`. I can do some reading and see if there is a more standard way to represent this though!

apdavison replied:
Looking into numpy's own typing, it looks like maybe we should use `import numpy.typing as npt` and 
```dtype: npt.DTypeLike = 'float32'```

zm711 replied:
I just put the same comment below :) haha.

apdavison replied:
sorry, just saw you posted the same thing below.

apdavison commented on neo/rawio/baserawio.py:
```diff
@@ -693,21 +725,23 @@ def get_event_timestamps(self, block_index=0, seg_index=0, event_channel_index=0
         return self._rescale_epoch_duration(raw_duration, dtype, event_channel_index)
 
-    def setup_cache(self, cache_path, **init_kargs):
+    def setup_cache(self, cache_path: 'home' | 'same_as_resource', **init_kargs):
```
shouldn't this be `cache_path: Literal['home', 'same_as_resource']` ?

apdavison commented on neo/rawio/baserawio.py:
```diff
@@ -735,7 +769,7 @@ def setup_cache(self, cache_path, **init_kargs):
         else:
             assert os.path.exists(cache_path), \
-                'cache_path do not exists use "home" or "same_as_resource" to make this auto'
+                'cache_path does not exists use "home" or "same_as_resource" to make this auto'
```
Not really related to the PR, but I don't understand what this error message is telling me. More punctutation and context needed.

zm711 replied:
Neither do I. But I think @samuelgarcia would have to comment on that.

apdavison reviewed (changes requested):
I am in favour of this PR, but some changes are needed for the type hints to be compatible with Python 3.8 and 3.9.

zm711 commented on neo/io/baseio.py:
```diff
@@ -10,6 +10,8 @@
 
 If you want a model for developing a new IO start from exampleIO.
 """
+from __future__ import annotations
```
This allows 3.8 and 3.9 to accept the |

zm711 commented on neo/rawio/baserawio.py:
```diff
@@ -67,6 +67,7 @@
 constructions of a RawIO for a given set of data.
 
 """
+from __future__ import annotations
```
same here this gives us 3.9 and 3.10 typing even in 3.8

zm711 commented:
> I am in favour of this PR, but some changes are needed for the type hints to be compatible with Python 3.8 and 3.9.

I think I addressed all of this in comments. But just to overall explain for future reference

I have added: 

```python
from __future__ import annotations
```

to the three files with typing, which allows us to use type hints from 3.9 and 3.10 in 3.8, 3.9, and 3.10 to keep things clean. This means we can:

* use `|` instead of needing `Literal`, but if `Literal` is preferred I can switch to that. 
* use built-in types even in 3.8 `list` instead of `List`, `dict` instead of `Dict`

apdavison commented:
> so if `|` was going to break the code we should've seen that during the testing.

yes, that confused me, but since you've explained about `from __future__ import annotations` it all makes sense!

There's one remaining point I'm not sure about (numpy.dtype), but otherwise this looks good.

zm711 commented:
Officially the better typing would be:

```python
import numpy.typing as npt

my_function(x, dtype: npt.DTypeLike = 'int32')
```

This indicates that something can be coerced into a numpy dtype. I currently have the dtype listed as the dtype constructor which is the constructor that takes an input and attempts to convert it to a numpy.dtype. So @apdavison you are correct that the current type is not strictly correct. Would you like me to change or do you think people will understand despite the mention of the constructor rather than the typing. I think the issue we will run into is that some old versions of numpy won't have the numpy typing module so it would be harder to support.

apdavison commented:
> I think the issue we will run into is that some old versions of numpy won't have the numpy typing module so it would be harder to support

That's a good point. For now we're not doing strict type checking with mypy or whatever, so I agree let's leave it as is for now.

zm711 commented:
Also just for reference I checked and typing was added in 1.21 so we would need to wait until our minimum numpy is that to be safe.
